### ISSUE TITLE
App Security Score inconsistencies

### ISSUE THREAD

**Issue by mat42290** [MobSF/Mobile-Security-Framework-MobSF#1940]
## ENVIRONMENT

```
OS and Version: all
Python Version: all
MobSF Version: 3.5.0
```

## EXPLANATION OF THE ISSUE

I'm noticing some inconsistencies in security scores since the release of 3.5.0.

I ran a static analysis on 2 applications and got these results :
![image](https://user-images.githubusercontent.com/43817651/162415048-06bbe682-4986-4680-b779-2981bd39d82b.png)
![image](https://user-images.githubusercontent.com/43817651/162415061-12bb3459-2e2c-4922-83a3-14ad09103578.png)

With this new security_score formula 
![image](https://user-images.githubusercontent.com/43817651/162415021-0417c966-9b0e-440b-9900-727ee4a45dad.png)
the security scores obtained are 36/100 and 39/100. So basically the first one which has way more findings receives a worse score ?

I understand the intention to ponder the score based on the ratio of the finding's criticity but that case seems to be a bit extreme.

**Comment by ajinabraham** [comment]
Thanks for brining this up. Do you have any thoughts on alternative implementations?

**Comment by ajinabraham** [comment]
Let me take a look at this.

**Comment by michaelkyawsan** [comment]
May I know, Is this issues solved in latest version?

**Comment by devtty1er** [comment]
@michaelkyawsan no, unfortunately the scoring system is really just a weighted average. E.g. if you have one high finding, it is "worse" than any number of high findings and some additional medium findings.

@ajinabraham please advise. Would you take a PR that reverts the scoring to pre-#1881?

> `mobsf/StaticAnalyzer/views/common/appsec.py`
```diff
high = len(findings.get('high'))
warn = len(findings.get('warning'))
sec = len(findings.get('secure'))
- total = high + warn + sec
- score = 0
- if total > 0:
-    score = int(100 - (
-        ((high * 1) + (warn * .5) - (sec * .2)) / total) * 100)
+ score = 100 - (high * 15) - (warn * 10) + (sec * 5)
if score > 100:
    score = 100
+ elif score < 0:
+   score = 10
findings['security_score'] = score

```

**Comment by johnxguo** [comment]
how about this algorithom
first,  calculate a "loss score"，e.g. , loss_score = high * 10 + warning * 5 - sec * 2，it‘s value range is (-∞,+∞), but in most case, the range is [0, +∞)
then, normalize the loss score to [-1, 1] or some other range, you can use softmax functions，e.g., sigmoid like functions
at last，map to your score interval
example code:
```
def sig_like(x):
    e = 2.71828
    return 2 / (1 + pow(e, x / 30))

def loss_score(high, warning, sec):
    return high * 10 + warning * 5  - sec * 2

def score(high, warning, sec):
    loss = loss_score(high, warning, sec)
    return min(sig_like(loss), 1) * 100

print(score(4, 3, 1))
```

**Comment by johnxguo** [comment]
> how about this algorithom first, calculate a "loss score"，e.g. , loss_score = high * 10 + warning * 5 - sec * 2，it‘s value range is (-∞,+∞), but in most case, the range is [0, +∞) then, normalize the loss score to [-1, 1] or some other range, you can use softmax functions，e.g., sigmoid like functions at last，map to your score interval example code:
> 
> ```
> def sig_like(x):
>     e = 2.71828
>     return 2 / (1 + pow(e, x / 30))
> 
> def loss_score(high, warning, sec):
>     return high * 10 + warning * 5  - sec * 2
> 
> def score(high, warning, sec):
>     loss = loss_score(high, warning, sec)
>     return min(sig_like(loss), 1) * 100
> 
> print(score(4, 3, 1))
> ```

@ajinabraham

**Comment by ajinabraham** [comment]
Do you want to send a PR?

**Comment by johnxguo** [comment]
> Do you want to send a PR?

#2311 @ajinabraham

**Comment by ajinabraham** [comment]
Thanks, I will test this out.

**Comment by sumit-jha-Pw** [comment]
I am also facing a similar issue where the score decreases even after improved results.
![Uploading Screenshot 2024-05-15 at 4.27.40 PM.png…]()

**Comment by sumit-jha-Pw** [comment]
I am also facing a similar issue where the score decreases even after improved results.
<img width="1419" alt="Screenshot 2024-05-15 at 4 27 40 PM" src="https://github.com/MobSF/Mobile-Security-Framework-MobSF/assets/147696560/1a0f7b1b-3b64-4d4e-af6a-d5c33b3c3d74">
![Uploading Screenshot 2024-05-15 at 4.27.40 PM.png…]()

**Comment by sumit-jha-Pw** [comment]
I am also facing a similar issue where the score decreases even after improved results.
<img width="1400" alt="Screenshot 2024-05-16 at 12 47 45 PM" src="https://github.com/MobSF/Mobile-Security-Framework-MobSF/assets/147696560/5eefa7f8-bbdd-448c-bc87-d1e5dcad7840">
![Uploading Screenshot 2024-05-15 at 4.26.16 PM.png…]()
